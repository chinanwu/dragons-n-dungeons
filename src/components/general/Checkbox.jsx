import React, { useState, useCallback } from 'react';
import PropTypes from 'prop-types';

import Input from './Input.jsx';

import './Checkbox.less';

export const Checkbox = ({ name, checked }) => {
  const [isChecked, setIsChecked] = useState(checked ? checked : false);

  const handleChange = useCallback(
    event => {
      setIsChecked(event.target.checked);
    },
    [setIsChecked]
  );

  return (
    <Input
      id="checkbox"
      className="Checkbox"
      name={name ? name : null}
      type="checkbox"
      checked={isChecked}
      onChange={handleChange}
    />
  );
};

Checkbox.propTypes = {
  name: PropTypes.string,
  checked: PropTypes.bool,
};

export default Checkbox;
