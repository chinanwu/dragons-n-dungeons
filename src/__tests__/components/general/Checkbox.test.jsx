import { shallow } from 'enzyme';
import React, { useState } from 'react';

import { Checkbox } from '../../../components/general/Checkbox.jsx';

jest.unmock('../../../components/general/Checkbox.jsx');

describe('Checkbox component', () => {
  describe('rendering', () => {
    it('renders Checkbox', () => {
      const wrapper = shallow(<Checkbox />);
      expect(wrapper.getElement()).toMatchSnapshot();
    });

    it('renders Checkbox with props', () => {
      const wrapper = shallow(<Checkbox name="test" checked={true} />);
      expect(wrapper.getElement()).toMatchSnapshot();
    });
  });

  describe('behaviour', () => {
    it('becomes checked when clicked', () => {
      const setIsChecked = jest.fn();
      useState.mockReturnValueOnce([false, setIsChecked]);
      const event = {
        target: {
          checked: true,
        },
      };
      const wrapper = shallow(<Checkbox />);
      expect(wrapper.getElement()).toMatchSnapshot();
      wrapper.find('#checkbox').simulate('change', event);
      expect(setIsChecked.mock.calls).toEqual([[true]]);
    });
  });
});
