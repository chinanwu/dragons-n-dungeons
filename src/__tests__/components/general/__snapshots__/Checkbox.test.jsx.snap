// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component behaviour becomes checked when clicked 1`] = `
<input
  checked={false}
  className="Checkbox"
  id="checkbox"
  name={null}
  onChange={[Function]}
  type="checkbox"
/>
`;

exports[`Checkbox component rendering renders Checkbox 1`] = `
<input
  checked={false}
  className="Checkbox"
  id="checkbox"
  name={null}
  onChange={[Function]}
  type="checkbox"
/>
`;

exports[`Checkbox component rendering renders Checkbox with props 1`] = `
<input
  checked={true}
  className="Checkbox"
  id="checkbox"
  name="test"
  onChange={[Function]}
  type="checkbox"
/>
`;
